### PR TITLE
Bidirectional attention flow

### PR DESCRIPTION
**Pull request by matt-gardner**
So far I just have the initial skeleton of the model here, but you can see how I'm thinking of doing this if you want to look at it now.  I don't expect to change the `forward()` method much from here, it's just that I need to implement a bunch of helper functions that I called that I haven't written yet.  It _should_ just be another day's work to implement those functions with decent tests, then the initial BiDAF implementation will be done.

### CODE REVIEW

**Comment by matt-gardner** [comment]
I will say that I like not having to wrap every operation in a `Layer` object (I'm assuming this actually works...).  And I like not having to think about how each of those operations should or shouldn't propagate a mask.  There are really only two masks you care about here, you get them once and use them in a few places, and you're done.  Masking isn't actually that complicated if you don't have to worry about propagating masks through layers everywhere.

File: allennlp/models/bidaf.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+        encoding_dim = phrase_layer.get_output_dim()
+        modeling_dim = modeling_layer.get_output_dim()
+        span_start_input_dim = encoding_dim * 4 + modeling_dim
+        self._span_start_predictor = TimeDistributed(torch.nn.Linear(span_start_input_dim, 1))
+
+        span_end_encoding_dim = span_end_encoder.get_output_dim()
+        span_end_input_dim = encoding_dim * 4 + span_end_encoding_dim
+        self._span_end_predictor = TimeDistributed(torch.nn.Linear(span_end_input_dim, 1))
+
+        # TODO(Mark): support masking once utility functions are merged.
```
Comment (DeNeutoy) [inline allennlp/models/bidaf.py]: haha cheeky!

Comment (matt-gardner) [inline allennlp/models/bidaf.py]: Copied directly from `simple_tagger.py` =).  You're the one who [knows what you meant](https://google.github.io/styleguide/pyguide.html?showone=TODO_Comments#TODO_Comments) there, so I kept your name.

**Comment by DeNeutoy** [comment]
Looks good - can you also implement an inference function like `SimpleTagger.tag` too? You were probably already going to do this, but just a reminder.

**Comment by matt-gardner** [comment]
Ok, @DeNeutoy, this should now be ready to review.

File: allennlp/models/bidaf.py [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+        return output_dict
+
+    def predict_span(self, question: TextField, passage: TextField) -> Dict[str, Any]:
+        """
+        Given a question and a passage, predicts the span in the passage that answers the question.
+
+        Parameters
+        ----------
+        question : ``TextField``
+        passage : ``TextField``
```
Comment (DeNeutoy) [inline allennlp/models/bidaf.py]: As this method is designed to be used in isolation of the `DatasetReader`, can you add the assumptions we make about the format of the passage, particularly that it needs to have a "@@STOP_TOKEN@@".

File: allennlp/models/bidaf.py [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+        The similarity function that we will use when comparing encoded passage and question
+        representations.
+    modeling_layer : ``Seq2SeqEncoder``
+        The encoder (with its own internal stacking) that we will use in between the bidirectional
+        attention and predicting span start and end.
+    span_end_encoder : ``Seq2SeqEncoder``
+        The encoder that we will use to incorporate span start predictions into the passage state
+        before predicting span end.
+    """
+
```
Comment (DeNeutoy) [inline allennlp/models/bidaf.py]: remove extra line here.

File: allennlp/models/bidaf.py [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+        self._span_start_predictor = TimeDistributed(torch.nn.Linear(span_start_input_dim, 1))
+
+        span_end_encoding_dim = span_end_encoder.get_output_dim()
+        span_end_input_dim = encoding_dim * 4 + span_end_encoding_dim
+        self._span_end_predictor = TimeDistributed(torch.nn.Linear(span_end_input_dim, 1))
+
+        # TODO(Mark): support masking once utility functions are merged.
+        self._loss = torch.nn.CrossEntropyLoss()
+
+        # TODO(mattg): figure out default initialization here
```
Comment (DeNeutoy) [inline allennlp/models/bidaf.py]: What needs to be done for this to happen? We have the classes for initalization already, is there something missing?

File: allennlp/models/bidaf.py [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+    def forward(self,  # type: ignore
+                question: Dict[str, torch.LongTensor],
+                passage: Dict[str, torch.LongTensor],
+                span_start: torch.IntTensor = None,
+                span_end: torch.IntTensor = None) -> Dict[str, torch.Tensor]:
+        # pylint: disable=arguments-differ
+        """
+        Parameters
+        ----------
+        question : Dict[str, torch.LongTensor]
```
Comment (DeNeutoy) [inline allennlp/models/bidaf.py]: These could be more informative, e.g that the passage must contain the answer span, we don't compute the loss  if the `span_start` and `span_end` aren't present. Also add full stops.

File: allennlp/models/bidaf.py [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+                                            dim=-1)
+        # Shape: (batch_size, passage_length, encoding_dim)
+        encoded_span_end = self._span_end_encoder(span_end_representation)
+        # Shape: (batch_size, passage_length, encoding_dim * 4 + span_end_encoding_dim)
+        span_end_input = torch.cat([final_merged_passage, encoded_span_end], dim=-1)
+        span_end_logits = self._span_end_predictor(span_end_input).squeeze(-1)
+        span_end_probs = masked_softmax(span_end_logits, passage_mask)
+
+        output_dict = {"span_start_logits": span_start_logits,
+                       "span_start_probs": span_start_probs,
```
Comment (DeNeutoy) [inline allennlp/models/bidaf.py]: Add `span_start_probs` and `span_end_probs` to the docstring.

File: allennlp/models/bidaf.py [hunk 7]
```diff
@@ -0,0 +1,348 @@
...
+        span_start_probs : numpy.ndarray
+        span_end_probs : numpy.ndarray
+        best_span : (int, int)
+        """
+        question.index(self._vocab)
+        passage.index(self._vocab)
+        # TODO(mattg): we should make the lengths an optional parameter to Instance.as_array()
+        question_lengths = question.get_padding_lengths()
+        question_input = arrays_to_variables(question.as_array(question_lengths))
+        for input_array in question_input.values():
```
Comment (DeNeutoy) [inline allennlp/models/bidaf.py]: Idea: We could add this functionality as a `assert_batch_dimension` flag to `arrays_as_variables`.

File: tests/common/tensor_test.py [hunk 8]
```diff
@@ -238,7 +238,7 @@ def test_last_dim_softmax_does_softmax_on_last_dim(self):
 
     def test_last_dim_softmax_handles_mask_correctly(self):
         batch_size = 1
-        length_1 = 5
```
Comment (DeNeutoy) [inline tests/common/tensor_test.py]: Why?

File: tests/models/bidaf_test.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+        self.token_indexers = {'tokens': SingleIdTokenIndexer(),
+                               'token_characters': TokenCharactersIndexer()}
+
+        self.model = BidirectionalAttentionFlow.from_params(self.vocab, Params({}))
+        initializer = InitializerApplicator()
+        initializer(self.model)
+
+    def test_forward_pass_runs_correctly(self):
+        training_arrays = arrays_to_variables(self.dataset.as_arrays())
+
```
Comment (DeNeutoy) [inline tests/models/bidaf_test.py]: Extra line here.

File: allennlp/models/bidaf.py (hunk 4 above)
Comment (matt-gardner) [inline allennlp/models/bidaf.py]: I'm not sure what the best way is to use them, or what I should be initializing things too.

File: allennlp/models/bidaf.py (hunk 7 above)
Comment (matt-gardner) [inline allennlp/models/bidaf.py]: Agreed that we should do something like this; I left out your TODO, but I'll put it back in.

File: tests/common/tensor_test.py (hunk 8 above)
Comment (matt-gardner) [inline tests/common/tensor_test.py]: I think this belonged with the prior tensor PR.  The reason was that there's an `expand_as` that works for the wrong reason if `num_options == length_1`, and it was masking an error that showed up in BiDAF.

Comment (matt-gardner) [inline tests/common/tensor_test.py]: I.e., this change made the test fail when I first made it.  It's not an issue anymore, because I fixed it in that other PR, but still probably better to have this changed.
